cdap-common/src/main/java/io/cdap/cdap/common/ApplicationNotFoundException.java

```diff
@@ -17,20 +17,27 @@
 package io.cdap.cdap.common;
 
 import io.cdap.cdap.proto.id.ApplicationId;
+import io.cdap.cdap.proto.id.ApplicationReference;
+import io.cdap.cdap.proto.id.NamespacedEntityId;
 
 /**
  * Thrown when an application is not found.
  */
 public class ApplicationNotFoundException extends NotFoundException {
 
-  private final ApplicationId id;
+  private final NamespacedEntityId id;
```

Comment:
hmmm, this change looks a bit confusing, especially that getId we return `NamespacedEntityId` instead of `ApplicationId`.

Maybe we should do something like this?

```
public class ApplicationNotFoundException extends NotFoundException {

  private final ApplicationId appId;
  private final ApplicationReference appRef;

  public ApplicationNotFoundException(ApplicationId id) {
    super(id);
    this.appId = id;
    this.appRef = id.getAppReference();
  }

  public ApplicationNotFoundException(ApplicationReference appReference) {
    super(appReference);
    this.appRef = appReference;
    this.appId = null;
  }

  public ApplicationId getId() {
    return appId;
  }

  public ApplicationReference getAppReference() {
    return appRef;
  }
}
```

Reply:
agree that it shouldn't be NamespacedEntityId, as that includes things that are not applications. Since you were able to remove the existing getId(), I guess that means it's not being used in a way where it needs to return an ApplicationId? If so it would be simpler to just have the ApplicaitonReference rather than both the id and reference.

Reply:
Yeah that could even more simplify the code @albertshau